### PR TITLE
fix(gridSystem.ts): call onEndReached when the data is less than cont…

### PR DESCRIPTION
**Pull request by prasannamestha**
This PR fixes this bug: https://github.com/petyosi/react-virtuoso/issues/1087

### CODE REVIEW

File: src/gridSystem.ts [hunk 1]
```diff
@@ -253,17 +253,21 @@ export const gridSystem = /*#__PURE__*/ u.system(
             } else {
               startIndex = perRow * floor((startOffset + rowGap) / (itemHeight + rowGap))
               endIndex = perRow * ceil((endOffset + rowGap) / (itemHeight + rowGap)) - 1
-              endIndex = min(totalCount - 1, max(endIndex, perRow - 1))
+              endIndex = min((data?.length || totalCount || 0) - 1, max(endIndex, perRow - 1))
```
Comment (petyosi) [inline src/gridSystem.ts:256]: Why is this change necessary?

File: src/gridSystem.ts [hunk 2]
```diff
@@ -272,9 +276,12 @@ export const gridSystem = /*#__PURE__*/ u.system(
...
       u.pipe(
-        data,
-        u.filter((data) => data !== null),
-        u.map((data) => data!.length)
+        u.combineLatest(data, totalCount),
+        u.filter(([data]) => data !== null),
+        u.map(([data, totalCount]) => {
+          return Math.max(data!.length, totalCount)
+        }),
+        u.distinctUntilChanged()
```
Comment (petyosi) [inline src/gridSystem.ts:284]: Why are you changing this? The component should accept either data or total count, and if data is supplied, it should control the total count value throughout the rest of the implementation.

File: src/gridSystem.ts [hunk 3]
```diff
@@ -253,17 +253,21 @@ export const gridSystem = /*#__PURE__*/ u.system(
...
+            // const rowCount = ceil(totalCount / perRow) // this is the total number of rows based on `totalCount`
+            const rowsVisibleCount = ceil((data?.length || totalCount || 0) / perRow) // number of rows rendered on screen
 
-            return { items, offsetTop: top, offsetBottom, top, bottom, itemHeight, itemWidth } as GridState
+            // const totalHeight = rowCount * itemHeight + (rowCount - 1) * rowGap // this is the total height that the container should occupy
+            const virtualHeight = rowsVisibleCount * itemHeight + (rowsVisibleCount - 1) * rowGap // height visible
+
+            const offsetBottom = virtualHeight - bottom
+
+            return { items, offsetTop: top, offsetBottom, top, bottom, itemHeight, itemWidth, data } as GridState
```
Comment (petyosi) [inline src/gridSystem.ts:270]: Unless I am missing something, this change should not do anything. See my comment on the relationship between totalCount and data.

File: src/gridSystem.ts [hunk 4]
```diff
@@ -310,11 +317,23 @@ export const gridSystem = /*#__PURE__*/ u.system(
...
+            (gridState.bottom !== 0 &&
+              gridState.itemHeight !== 0 &&
+              gridState.offsetBottom === 0 &&
+              gridState.items.length === data?.length &&
+              // gridState.items[gridState.items.length - 1].index < totalCount - 1
+              gridState.items[gridState.items.length - 1].index === (data?.length || 0) - 1)
+          )
+        }),
+        u.map(([[_, data]]) => {
+          return data?.length
```
Comment (petyosi) [inline src/gridSystem.ts:335]: How does that work if the grid has no data and uses totalCount?

File: src/gridSystem.ts (hunk 1 above)
Comment (prasannamestha) [reply src/gridSystem.ts:256]: I'll prepare a video later today to explain but here's the gist:

When both `data` and `totalCount` is passed to `VirtuosoGrid` - the endIndex calculation was faulty. Now looking at this after a good night sleep, it seems like we might not require this as we update `totalCount` below. I'll get back on this

File: src/gridSystem.ts (hunk 3 above)
Comment (prasannamestha) [reply src/gridSystem.ts:270]: My bad, I added `data` here for testing, forgot removing it. Will update this later today

File: src/gridSystem.ts (hunk 2 above)
Comment (prasannamestha) [reply src/gridSystem.ts:284]: But it is a common use case for something like a dynamic list (or infinite list)

File: src/gridSystem.ts (hunk 1 above)
Comment (prasannamestha) [reply src/gridSystem.ts:256]: Okay here's the explanation why this change is necessary:

In a component with paginated data (data received from `useInfiniteQuery`), `onEndReached` function fires the callback to load more data - that eventually increases the data size. Additionally there might be a different `totalIndex` specified to calculate the container height. So usually `totalCount` could be greater than or equal to `data.length`

Hence the `endIndex` should consider the min of `data.length` or `totalCount` to get the right value

File: src/gridSystem.ts (hunk 2 above)
Comment (petyosi) [reply src/gridSystem.ts:284]: I can see how one can potentially achieve some elaborate implementation using a combination of both props and load on demand, and conditional rendering depending on data being present or not.

However, this is way beyond trivial, and supporting this would need a good set of examples and tests. If these are not present, the capability would just result in people struggling infinitely and claiming "bugs" that I need to deal with.

Comment (prasannamestha) [reply src/gridSystem.ts:284]: I see.. Gotcha, maybe I touched a bit more code than necessary. I'll make some changes

Comment (prasannamestha) [reply src/gridSystem.ts]: @petyosi take a look now, I have added an example page and updated code to only touch `endReached` fn.

File: src/gridSystem.ts [hunk 5]
```diff
@@ -310,11 +310,25 @@ export const gridSystem = /*#__PURE__*/ u.system(
...
+        u.filter(([[gridState, data], _totalCount, hasScrolled]) => {
+          const lastIndex = gridState.items[gridState.items.length - 1].index
+          const isLastItemRendered = lastIndex === (data?.length || 0) - 1
+
+          // User has scrolled
+          if (hasScrolled) return isLastItemRendered
+
+          // User has not scrolled, so check whether grid is fully rendered
+          const isFullyRendered =
+            gridState.bottom > 0 && gridState.itemHeight > 0 && gridState.offsetBottom === 0 && gridState.items.length === data?.length
```
[the remark this answers is not in the document]
Comment (prasannamestha) [reply src/gridSystem.ts:325]: It's strange but using `totalCount` here doesn't fire more than once. But I think I found a way to get around it. Hold on, sending an update

[the remark this answers is not in the document]
Comment (prasannamestha) [reply src/gridSystem.ts]: @petyosi check now :)
